crates/common/consensus/lean/src/state.rs

```diff
@@ -48,126 +60,373 @@ impl LeanState {
+                .get(root)
+                .ok_or_else(|| anyhow!("Root {root} not found in justifications"))?;
+
+            // Assert that votes list has exactly VALIDATOR_REGISTRY_LIMIT items.
+            // If the length is incorrect, the constructed bitlist will be corrupt.
+            ensure!(
+                justifications_for_root.len() == VALIDATOR_REGISTRY_LIMIT as usize,
+                "Justifications length does not match validator registry limit"
+            );
+
+            justifications_roots
+                .push(*root)
+                .map_err(|err| anyhow!("Failed to add root to justifications_roots: {err:?}"))?;
+
+            justifications_for_root
+                .iter()
+                .for_each(|justification| flattened_justifications.push(justification));
         }
 
-        self.justifications_roots
-            .push(*root)
-            .map_err(|err| anyhow!("Failed to insert root into justifications_roots: {err:?}"))?;
+        // Create a new Bitlist with all the flattened votes
+        let mut justifications_validators = BitList::with_capacity(
+            (MAX_HISTORICAL_BLOCK_HASHES * VALIDATOR_REGISTRY_LIMIT) as usize,
+        )
+        .map_err(|err| {
+            anyhow!("Failed to create BitList for justifications_validators: {err:?}")
+        })?;
 
-        let old_length = self.justifications_roots_validators.len();
-        let new_length = old_length + VALIDATOR_REGISTRY_LIMIT as usize;
+        flattened_justifications.iter().enumerate().try_for_each(
+            |(index, justification)| -> anyhow::Result<()> {
+                justifications_validators
+                    .set(index, *justification)
+                    .map_err(|err| anyhow!("Failed to set justification bit: {err:?}"))
+            },
+        )?;
 
-        let mut new_justifications_roots_validators = BitList::with_capacity(new_length)
-            .map_err(|err| anyhow!("Failed to initialize new justification bits: {err:?}"))?;
+        self.justifications_roots = justifications_roots;
+        self.justifications_validators = justifications_validators;
 
-        for (i, bit) in self.justifications_roots_validators.iter().enumerate() {
-            new_justifications_roots_validators
-                .set(i, bit)
-                .map_err(|err| {
-                    anyhow!("Failed to initialize justification bits to existing values: {err:?}")
-                })?;
+        Ok(())
+    }
+
+    pub fn state_transition(
+        &mut self,
+        signed_block: &SignedBlock,
+        valid_signatures: bool,
+        validate_result: bool,
+    ) -> anyhow::Result<()> {
+        // Verify signatures
+        ensure!(valid_signatures, "Signatures are not valid");
+
+        let block = &signed_block.message;
+
+        // Process slots (including those with no blocks) since block
+        self.process_slots(block.slot)
+            .context("Failed to process slots")?;
```

Comment:
nice anyhow find (not sure if it was Jun or you) but this is def clean if we are using an anyhow result